I have tried to enforce the 64-bit integers on the index arrays.
```
def test_interval_index_reindex_behavior(base, expected_result):
    # GH 51826
    left = np.arange(base, dtype=np.int64)
    right = np.arange(1, base + 1, dtype=np.int64)
    d = Series(
        range(base),
        index=IntervalIndex.from_arrays(left, right)
    )
    result = d.reindex(index=[np.nan, 1.0])
    tm.assert_series_equal(result, expected_result)
 ```
 However, the 32 bit tests still seems to fail. The specific error is 
 ```
 FAILED pandas/tests/indexing/interval/test_interval.py::TestIntervalIndexInsideMultiIndex::test_interval_index_reindex_behavior[101-expected_result2] - TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
FAILED pandas/tests/indexing/interval/test_interval.py::TestIntervalIndexInsideMultiIndex::test_interval_index_reindex_behavior[1010-expected_result3] - TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
```

And the error also points to `line:196` on `pandas/_libs/intervaltree.pxi` which is this return statement.
```
cdef take(ndarray source, ndarray indices):
    """Take the given positions from a 1D ndarray
    """
    return PyArray_Take(source, indices, 0)
```
 Do anyone have ideas on solving this unexpected casting issue?

--- a/pandas/tests/indexing/interval/test_interval.py
+++ b/pandas/tests/indexing/interval/test_interval.py
@@ -1,6 +1,8 @@
 import numpy as np
 import pytest
 
+from pandas.compat import IS64
+
 import pandas as pd
 from pandas import (
     DataFrame,
@@ -173,15 +175,18 @@ class TestIntervalIndexInsideMultiIndex:
         expected = Series([1, 6, 2, 8, 7], index=expected_index, name="value")
         tm.assert_series_equal(result, expected)
 
+    @pytest.mark.xfail(not IS64, reason="GH 23440")
     @pytest.mark.parametrize(
         "base",
-        [10, 100, 101, 1010],
+        [101, 1010],
     )
     def test_reindex_behavior_with_interval_index(self, base):
         # GH 51826
-        left = np.arange(base)
-        right = np.arange(1, base + 1)
-        ser = Series(range(base), index=IntervalIndex.from_arrays(left, right))
-        expected_result = Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)
+
+        ser = Series(
+            range(base),
+            index=IntervalIndex.from_arrays(range(base), range(1, base + 1)),
+        )
+        expected_result = Series([np.nan, 0], index=[np.nan, 1.0], dtype=float)
         result = ser.reindex(index=[np.nan, 1.0])
         tm.assert_series_equal(result, expected_result)

--- a/pandas/tests/indexing/interval/test_interval.py
+++ b/pandas/tests/indexing/interval/test_interval.py
@@ -174,18 +174,14 @@ class TestIntervalIndexInsideMultiIndex:
         tm.assert_series_equal(result, expected)
 
     @pytest.mark.parametrize(
-        "base, expected_result",
-        [
-            (10, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
-            (100, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
-            (101, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
-            (1010, Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)),
-        ],
+        "base",
+        [10, 100, 101, 1010],
     )
-    def test_reindex_behavior_with_interval_index(self, base, expected_result):
+    def test_reindex_behavior_with_interval_index(self, base):
         # GH 51826
         left = np.arange(base)
         right = np.arange(1, base + 1)
-        d = Series(range(base), index=IntervalIndex.from_arrays(left, right))
-        result = d.reindex(index=[np.nan, 1.0])
+        ser = Series(range(base), index=IntervalIndex.from_arrays(left, right))
+        expected_result = Series([np.nan, 0], index=[np.nan, 1.0], dtype=np.float64)
+        result = ser.reindex(index=[np.nan, 1.0])
         tm.assert_series_equal(result, expected_result)

